experimental/hip/memory_pools.c

```diff
@@ -0,0 +1,326 @@
+                                           IREE_HAL_MEMORY_TYPE_DEVICE_LOCAL);
+  (void)is_device_local;
+  iree_device_size_t allocation_size = iree_hal_buffer_allocation_size(buffer);
+  (void)allocation_size;
+  IREE_TRACE_ALLOC_NAMED(
+      is_device_local ? IREE_HAL_HIP_DEVICE_LOCAL_POOL_RESERVED_ID
+                      : IREE_HAL_HIP_OTHER_POOL_RESERVED_ID,
+      (void*)iree_hal_hip_buffer_device_pointer(buffer), allocation_size);
+  IREE_STATISTICS({
+    iree_atomic_int64_t* bytes_allocated =
+        is_device_local ? &pools->statistics.device_bytes_allocated
+                        : &pools->statistics.host_bytes_allocated;
+    iree_atomic_fetch_add_int64(bytes_allocated, allocation_size,
+                                iree_memory_order_relaxed);
+  });
+}
+
+static void iree_hal_hip_memory_pool_track_free(
+    iree_hal_hip_memory_pools_t* pools, iree_hal_buffer_t* buffer) {
+  bool is_device_local = iree_all_bits_set(iree_hal_buffer_memory_type(buffer),
+                                           IREE_HAL_MEMORY_TYPE_DEVICE_LOCAL);
+  (void)is_device_local;
+  IREE_TRACE_FREE_NAMED(is_device_local
+                            ? IREE_HAL_HIP_DEVICE_LOCAL_POOL_RESERVED_ID
+                            : IREE_HAL_HIP_OTHER_POOL_RESERVED_ID,
+                        (void*)iree_hal_hip_buffer_device_pointer(buffer));
+  IREE_STATISTICS({
+    iree_atomic_int64_t* bytes_freed =
+        is_device_local ? &pools->statistics.device_bytes_freed
+                        : &pools->statistics.host_bytes_freed;
+    iree_device_size_t allocation_size =
+        iree_hal_buffer_allocation_size(buffer);
+    iree_atomic_fetch_add_int64(bytes_freed, allocation_size,
+                                iree_memory_order_relaxed);
+  });
+}
+
+void iree_hal_hip_memory_pools_merge_statistics(
+    iree_hal_hip_memory_pools_t* pools,
+    iree_hal_allocator_statistics_t* statistics) {
+  IREE_STATISTICS({
+    statistics->device_bytes_allocated = iree_atomic_load_int64(
+        &pools->statistics.device_bytes_allocated, iree_memory_order_relaxed);
+    statistics->host_bytes_allocated = iree_atomic_load_int64(
+        &pools->statistics.host_bytes_allocated, iree_memory_order_relaxed);
+    statistics->device_bytes_freed = iree_atomic_load_int64(
+        &pools->statistics.device_bytes_freed, iree_memory_order_relaxed);
+    statistics->host_bytes_freed = iree_atomic_load_int64(
+        &pools->statistics.host_bytes_freed, iree_memory_order_relaxed);
+
+    // WARNING: hipMemPoolGetAttribute() API is marked as beta in HIP library
+    // meaning that while the feature is complete, it is still open to changes
+    // and may have outstanding issues.
+    if (pools->device_local) {
+      uint64_t pool_peak = 0;
+      IREE_HIP_IGNORE_ERROR(
+          pools->hip_symbols,
+          hipMemPoolGetAttribute(pools->device_local,
+                                 hipMemPoolAttrReservedMemHigh, &pool_peak));
```

Comment:
This should be [hipMemPoolAttrUsedMemHigh](https://docs.amd.com/projects/HIP/en/latest/.doxygen/docBin/html/group___global_defs.html#gga987c8e7a7e8171832a6647150854ca2ea7d0d0e5e6ce19fdaaef2674a305930d0)?